File: 116.填充每个节点的下一个右侧节点指针.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */
/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function (root) {
    if (root == null) {
        return root
    }
    let queue = []
    queue.push(root)
    while (queue.length != 0) {
        const length = queue.length
        for (let i = 0; i < length; i++) {
            const prev = queue[i]
            if (i != length - 1) {
                prev.next = queue[i + 1]
            } else {
                prev.next = null
            }
            if (prev.left) {
                queue.push(prev.left)
            }
            if (prev.right) {
                queue.push(prev.right)
            }
        }
        // console.log(length, queue)
        queue.splice(0, length)
        // console.log(length, queue)
    }
    return root
};

File: 116.填充每个节点的下一个右侧节点指针.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	d := make(map[int]*Node)
	var dfs func(node *Node, depth int)
	dfs = func(node *Node, depth int) {
		if node == nil {
			return
		}
		if prev, ok := d[depth]; ok {
			prev.Next = node
			d[depth] = node
		} else {
			d[depth] = node
		}
		dfs(node.Left, depth+1)
		dfs(node.Right, depth+1)
	}
	dfs(root, 0)
	return root
}